core/src/main/java/org/apache/iceberg/SerializableTableFactory.java

```diff
@@ -0,0 +1,383 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.hadoop.HadoopFileIO;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.SerializableMap;
+
+/**
+ * A factory to create serializable tables.
+ */
+public class SerializableTableFactory {
```

Comment:
I went with a factory instead of a util class so that I can have it in `org.apache.iceberg` and don't need to open up our base metadata table class.

Reply:
this doesn't look like a traditional factory class. We probably don't need this wrapper and directly make the `SerializableTable` a public class.

Then we can either have a constructor like `SerializableTable(Table t)` or have a static `public staticSerializableTable#from(Table t)`

Reply:
We have two table implementations: one for base and one for metadata tables. I don't think making those two classes public is a good idea. We need to do a switch somewhere and expose a method for building serializable tables.

While we could name the class like `SerializableTable` and do a switch there, it may be a bit weird that it does not implement `Table`. Let me think more about this. Maybe, that's fine.

```
SerializableTable.copyOf(table)
SerializableTables.copyOf(table)
```

Reply:
`SerializableTables.copyOf(table)` looks more accurate than `Factory` to me.

Reply:
I've updated to `SerializableTable.copyOf(table)` and made the metadata table a private nested class.
Could you take one more look, @stevenzwu?